### ISSUE TITLE
Discussion: More immutability in webnative

### ISSUE THREAD

**Issue by matheus23** [fission-codes/webnative#226]
(Came up in a call with @expede)

Recently a couple of concurrency issues surfaced. They stem from data races, so the root problem in these cases is that some *mutable* state is modified by multiple entities.

We should think about making more of our data structures immutable.

Concretely, this would mean e.g. making the `MMPT` class only have immutable members, and e.g. its `add()` method return a new reference to a new MMPT.

We also have data races - as the `concurrency.test.ts` sometimes reveals - in the private file system code in (*I think*) `PrivateTree` (I did some testing in the `matheus23/fix-concurrency`).

We should think about also making this data structure immutable.

---

This is obviously a bigger refactor, so we should plan this a little better. Maybe we do this incrementally?

Also, at some point we have to put a mutable API over the immutable code, as that's more in line of how e.g. an app developer thinks about data storage (duh!).

We think we might want to have a very thin "mutability" layer, think of it like an index into the individual, immutable private file system trees

I'm thinking of also exposing the immutable file system stuff, so it's still composable, because once you create a "mutable" layer on top of an immutable API, you'll never truly get back the immutable API. I don't want to lose the immutable API in case anyone wants to integrate it into an immutable-first codebase.

**Comment by expede** [comment]
> This is obviously a bigger refactor, so we should plan this a little better

Yeah, this is totally a "measure twice / cut once".

I've been doing some poking around. You're going to laugh when you see the state of the art: "In this paper we use optimistic concurrency techniques adapted from software transactional memory to develop a concurrent tree data structure." Soooo exactly what we were saying on the call 😆 Except that we're _implementing_ something STM-like, so that doesn't really help us too much.

> We think we might want to have a very thin "mutability" layer, think of it like an index into the individual, immutable private file system trees

Yeah, exactly. There's a few edge cases, but in theory a light reference index for the latest version feels like the right direction.

> Maybe we do this incrementally?

That would be ideal. Let's get this fully specced out, and then see which bit can be done first 👍 Okasaki was surprisingly unhelpful on concurrent inserts, aside from the classic merge operation which doesn't totally work for use because we have namespace contention. I have some papers open; will share any findings!

**Comment by expede** [comment]
Working thoughts on approaches:

Switching to a less mutable solution makes keeping track of various parts of the state much more straightforward, _and_ we get to align with the immutability inherent in content addressing.

# Write-Ahead Log

* Fairly straightforward to implement
* Doesn't solve for _realtime_ FS concurrency, does batches
* Difficult to guarantee consistency between remote `HEAD` and local state
  * Can trade off performance to get consistency

# Global Rebuilding

* Per Okasaki ~p102, Hood & Melville realtime queues)
* Separation of working copy and queryable (finalized) structure
  * AKA persistent version of the classic clean and dirty states
    * Slightly cleaner than in the mutable case, because persistence
  * Leave in a `Promise` until available in finalized version?
* Okasaki warns that these tend to be very difficult to implement correctly
  * ...for the same reasons that the our current implementation is suffering
    * (i.e. handlig multiple concurrent edits of the same node)

# Immutable Data / Mutable Spine

* Essentially an STM index pointing at an immutable store
* Doubles as a snapshot of the current `HEAD`

**Comment by matheus23** [comment]
I find the distinction between fully and mostly immutable confusing.

If the surface level API is supposed to look like this:

```js
const permissions = ...
const { fs } = await webnative.initialise({ permissions })
fs.write("private/a.txt", "sth")
fs.write("private/b.txt", "sth")
```

Then we will always have to have a layer of mutability in there.

The only way we can ever be fully immutable is when we expose an API like this:

```js
const permissions = ...
const { fs } = await webnative.initialise({ permissions })
const fs2 = fs.write("private/a.txt", "sth")
const fs3 = fs2.write("private/b.txt", "sth")
```

(At least that's how I am thinking about this in my head)

I must be misunderstanding what you refer to when you/Okazaki says "can be implemented fully immutable".

Also, I admit I haven't read the okazaki book. I should really do that :)

**Comment by matheus23** [comment]
Ok, so I've been chewing on this for a bit.

Given that we need some kind of merging operation on filesystems when we implement conflict resolution (rebasing), I think it makes a lot of sense to implement the optimistic fork/merge strategy.

This reminds me that this discussion is probably also highly relevant for filesystem transactions in webnative. (isn't a transaction basically a fork & merge?)

**Comment by expede** [comment]
> This reminds me that this discussion is probably also highly relevant for filesystem transactions in webnative. (isn't a transaction basically a fork & merge?)

They're closely related! The atomic transactions are higher granularity than this: literally toss the existing tree, do a rebase (with the same merge) and replay the anonymous function(s).

Can we make this one mechanism? Probably! Needs a bit more thought, though.

**Comment by expede** [comment]
(also FWIW I also have a gut reaction in favour of the fork/merge option)

**Comment by expede** [comment]
Random thought while in an unrelated call, so just a quick sketch of a stray idea:

In fork/merge, if we're doing associative merges with history, we don't want to have to walk back in the linked list to fix all previous references when there's no contention.

It may make sense to do this with CPS while in the array, to make merging to an arbitrary spot more performant (where the continuation is the part pointing back at unknown history).

**Comment by icidasset** [comment]
Closed by #500
Part of #501
